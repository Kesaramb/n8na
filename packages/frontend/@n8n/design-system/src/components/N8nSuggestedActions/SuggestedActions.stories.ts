import type { StoryFn } from '@storybook/vue3';

import N8nSuggestedActions from './SuggestedActions.vue';

export default {
	title: 'Modules/SuggestedActions',
	component: N8nSuggestedActions,
	argTypes: {},
};

const Template: StoryFn = (args, { argTypes }) => ({
	setup() {
		return { args };
	},
	props: Object.keys(argTypes),
	components: {
		N8nSuggestedActions,
	},
	template: `
		<div style="padding: 50px;">
			<N8nSuggestedActions v-bind="args" @action-click="onActionClick" @ignore-click="onIgnoreClick" />
		</div>
	`,
	methods: {
		onActionClick(actionId: string) {
			console.log('Action clicked:', actionId);
			alert(`Action clicked: ${actionId}`);
		},
		onIgnoreClick(actionId: string) {
			console.log('Ignore clicked:', actionId);
			alert(`Ignore clicked: ${actionId}`);
		},
	},
});

export const Default = Template.bind({});
Default.args = {
	actions: [
		{
			id: 'evaluate-workflow',
			title: 'Evaluate your workflow with a dataset',
			description: 'Set up an AI evaluation to be sure th WF is reliable.',
			moreInfoLink: 'https://docs.n8n.io/evaluations',
			buttonLabel: 'Go to evaluations',
		},
		{
			id: 'track-errors',
			title: 'Keep track of execution errors',
			description: 'Setup a workflow error to track what is going on here.',
			moreInfoLink: 'https://docs.n8n.io/error-workflows',
			buttonLabel: 'Settings',
		},
		{
			id: 'track-time',
			title: 'Track the time you save',
			description:
				'Log how much time this workflow saves each run to measure your automation impact.',
			buttonLabel: 'Settings',
		},
	],
};

export const WithoutMoreInfoLinks = Template.bind({});
WithoutMoreInfoLinks.args = {
	actions: [
		{
			id: 'action-1',
			title: 'Action without more info link',
			description: 'This action does not have a more info link.',
			buttonLabel: 'Do something',
		},
		{
			id: 'action-2',
			title: 'Another action',
			description: 'This one also lacks a more info link.',
			buttonLabel: 'Do another thing',
		},
	],
};

export const LongContent = Template.bind({});
LongContent.args = {
	actions: Array.from({ length: 10 }, (_, i) => ({
		id: `action-${i + 1}`,
		title: `Suggested Action ${i + 1}`,
		description: `This is a longer description for action ${i + 1} that demonstrates how the component handles scrolling when there are many items in the list.`,
		moreInfoLink: i % 2 === 0 ? 'https://docs.n8n.io' : undefined,
		buttonLabel: `Action ${i + 1}`,
	})),
};

const TemplateWithEvents: StoryFn = (args, { argTypes }) => ({
	setup() {
		return { args };
	},
	props: Object.keys(argTypes),
	components: {
		N8nSuggestedActions,
	},
	template: `
		<div style="padding: 50px;">
			<N8nSuggestedActions
				v-bind="args"
				@action-click="onActionClick"
				@ignore-click="onIgnoreClick"
				@ignore-all="onIgnoreAll"
			/>
		</div>
	`,
	methods: {
		onActionClick(actionId: string) {
			console.log('Action clicked:', actionId);
			alert(`Action clicked: ${actionId}`);
		},
		onIgnoreClick(actionId: string) {
			console.log('Ignore clicked:', actionId);
			alert(`Ignore clicked: ${actionId}`);
		},
		onIgnoreAll() {
			console.log('Ignore all clicked');
			alert('Ignore all clicked');
		},
	},
});

export const WithIgnoreAllOption = TemplateWithEvents.bind({});
WithIgnoreAllOption.args = {
	ignoreAllLabel: 'Ignore for all workflows',
	actions: [
		{
			id: 'evaluate-workflow',
			title: 'Evaluate your workflow with a dataset',
			description: 'Set up an AI evaluation to be sure th WF is reliable.',
			moreInfoLink: 'https://docs.n8n.io/evaluations',
			buttonLabel: 'Go to evaluations',
		},
		{
			id: 'track-errors',
			title: 'Keep track of execution errors',
			description: 'Setup a workflow error to track what is going on here.',
			moreInfoLink: 'https://docs.n8n.io/error-workflows',
			buttonLabel: 'Settings',
		},
	],
};

export const SingleActionWithTurnOff = TemplateWithEvents.bind({});
SingleActionWithTurnOff.args = {
	ignoreAllLabel: 'Disable all suggestions',
	actions: [
		{
			id: 'single-action',
			title: 'Single action with turn off option',
			description: 'This shows how the turn off option appears even with a single action.',
			buttonLabel: 'Take action',
		},
	],
};

const AlignmentTemplate: StoryFn = (args, { argTypes }) => ({
	setup() {
		return { args };
	},
	props: Object.keys(argTypes),
	components: {
		N8nSuggestedActions,
	},
	template: `
		<div style="padding: 50px; display: flex; justify-content: space-between; width: 800px;">
			<div>
				<h4 style="margin-bottom: 10px;">Start Alignment</h4>
				<N8nSuggestedActions v-bind="args" popoverAlignment="start" />
			</div>
			<div>
				<h4 style="margin-bottom: 10px;">Center Alignment</h4>
				<N8nSuggestedActions v-bind="args" popoverAlignment="center" />
			</div>
			<div>
				<h4 style="margin-bottom: 10px;">End Alignment</h4>
				<N8nSuggestedActions v-bind="args" popoverAlignment="end" />
			</div>
		</div>
	`,
});

export const PopoverAlignments = AlignmentTemplate.bind({});
PopoverAlignments.args = {
	actions: [
		{
			id: 'action1',
			title: 'Test Alignment',
			description: 'This demonstrates different popover alignments.',
			buttonLabel: 'Action',
		},
	],
};

export const MultipleActionsWithIgnoreAll = TemplateWithEvents.bind({});
MultipleActionsWithIgnoreAll.args = {
	actions: [
		{
			id: 'action1',
			title: 'First action',
			description: 'This is the first action that can be ignored.',
			buttonLabel: 'Do first thing',
			completed: true,
		},
		{
			id: 'action2',
			title: 'Second action',
			description: 'This is the second action that can be ignored.',
			buttonLabel: 'Do second thing',
		},
		{
			id: 'action3',
			title: 'Third action',
			description: 'This is the third action that can be ignored.',
			buttonLabel: 'Do third thing',
		},
	],
};
