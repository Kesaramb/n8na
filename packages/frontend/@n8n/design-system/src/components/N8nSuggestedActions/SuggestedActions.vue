<script setup lang="ts">
import { computed, ref } from 'vue';

import { useI18n } from '../../composables/useI18n';
import N8nHeading from '../N8nHeading';
import N8nIcon from '../N8nIcon';
import N8nLink from '../N8nLink';
import N8nPopoverReka from '../N8nPopoverReka';
import N8nTag from '../N8nTag';
import N8nText from '../N8nText';

interface SuggestedAction {
	id: string;
	title: string;
	description: string;
	moreInfoLink?: string;
	completed?: boolean;
}

export interface SuggestedActionsProps {
	title: string;
	actions: SuggestedAction[];
	open: boolean;
	ignoreAllLabel?: string;
	popoverAlignment?: 'start' | 'end' | 'center';
}

interface SuggestedActionsEmits {
	(event: 'action-click', actionId: string): void;
	(event: 'ignore-click', actionId: string): void;
	(event: 'ignore-all'): void;
	(event: 'update:open', open: boolean): void;
}

defineOptions({ name: 'N8nSuggestedActions' });

const props = withDefaults(defineProps<SuggestedActionsProps>(), {
	ignoreAllLabel: undefined,
	popoverAlignment: undefined,
});

const emit = defineEmits<SuggestedActionsEmits>();
const { t } = useI18n();

const ignoringActions = ref<Set<string>>(new Set());

const completedCount = computed(() => props.actions.filter((action) => action.completed).length);

const handleActionClick = (action: SuggestedAction) => {
	if (!action.completed) {
		emit('action-click', action.id);
	}
};

const handleIgnoreClick = (actionId: string) => {
	ignoringActions.value.add(actionId);
	setTimeout(() => {
		emit('ignore-click', actionId);
		ignoringActions.value.delete(actionId);
	}, 500);
};
</script>

<template>
	<N8nPopoverReka
		v-if="completedCount !== actions.length"
		:open="open"
		width="360px"
		max-height="500px"
		:align="popoverAlignment"
		@update:open="$emit('update:open', $event)"
	>
		<template #trigger>
			<div :class="$style.triggerContainer" data-test-id="suggested-action-count">
				<N8nTag :text="`${completedCount} / ${actions.length}`" />
			</div>
		</template>
		<template #content>
			<div :class="$style.popoverContent">
				<N8nHeading tag="h4">{{ title }}</N8nHeading>
				<div
					v-for="action in actions"
					:key="action.id"
					:class="[
						{
							[$style.actionItem]: true,
							[$style.ignoring]: ignoringActions.has(action.id),
							[$style.actionable]: !action.completed,
						},
					]"
					data-test-id="suggested-action-item"
					:data-action-id="action.id"
					@click.prevent.stop="() => handleActionClick(action)"
				>
					<div :class="$style.checkboxContainer">
						<N8nIcon v-if="action.completed" icon="circle-check" color="success" />
						<N8nIcon v-else icon="circle" color="foreground-dark" />
					</div>
					<div :class="$style.actionItemBody">
						<div :class="[action.completed ? '' : 'mb-3xs', $style.actionHeader]">
							<N8nText size="medium" :bold="true">{{ action.title }}</N8nText>

							<N8nIcon v-if="!action.completed" icon="chevron-right" />
						</div>
						<div v-if="!action.completed">
							<N8nText size="small" color="text-base">
								{{ action.description }}
								<N8nLink
									v-if="action.moreInfoLink"
									:to="action.moreInfoLink"
									size="small"
									theme="text"
									new-window
									underline
									@click.stop
								>
									{{ t('generic.moreInfo') }}
								</N8nLink>
							</N8nText>
						</div>
						<div v-if="!action.completed" :class="$style.actionButtons">
							<N8nLink
								theme="text"
								size="small"
								data-test-id="suggested-action-ignore"
								underline
								@click.prevent.stop="handleIgnoreClick(action.id)"
							>
								{{ t('generic.ignore') }}
							</N8nLink>
						</div>
					</div>
				</div>
				<div :class="$style.ignoreAllContainer">
					<N8nLink
						theme="text"
						size="small"
						underline
						data-test-id="suggested-action-ignore-all"
						@click.prevent.stop="emit('ignore-all')"
					>
						{{ ignoreAllLabel ?? t('generic.ignoreAll') }}
					</N8nLink>
				</div>
			</div>
		</template>
	</N8nPopoverReka>
</template>

<style lang="scss" module>
.triggerContainer {
	display: inline-block;
	position: relative;
	--tag-height: 24px;
}

.popoverContent {
	display: flex;
	flex-direction: column;
	gap: var(--spacing-s);
	padding: var(--spacing-m) var(--spacing-s);
}

.actionItem {
	display: flex;
	flex-direction: row;
	transition:
		opacity 0.3s ease,
		filter 0.3s ease;

	&.ignoring {
		opacity: 0.5;
		filter: grayscale(0.8);
		pointer-events: none;
		cursor: not-allowed;
	}
}

.actionable {
	&:hover {
		cursor: pointer;

		.actionHeader {
			color: var(--color-primary);
		}

		&:has(a:hover) {
			.actionHeader {
				color: var(--color-text-dark);
			}
		}
	}
}

.actionItemBody {
	display: flex;
	flex-direction: column;
	flex-grow: 1;
	border-bottom: var(--border-base);
	padding-bottom: var(--spacing-s);
}

.checkboxContainer {
	padding-top: 1px;
	padding-right: var(--spacing-xs);
}

.actionHeader {
	display: flex;
	align-items: flex-start;

	> :first-child {
		flex-grow: 1;
	}
}

.actionButtons {
	margin-top: var(--spacing-2xs);
}

.ignoreAllContainer {
	padding-left: var(--spacing-l);
}
</style>
