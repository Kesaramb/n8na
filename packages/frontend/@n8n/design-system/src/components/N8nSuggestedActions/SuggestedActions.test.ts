import { render, fireEvent } from '@testing-library/vue';
import { describe, it, expect } from 'vitest';

import N8nSuggestedActions from './SuggestedActions.vue';

const mockActions = [
	{
		id: 'action1',
		title: 'Evaluate your workflow',
		description: 'Set up an AI evaluation to be sure th WF is reliable.',
		moreInfoLink: 'https://docs.n8n.io/evaluations',
	},
	{
		id: 'action2',
		title: 'Keep track of execution errors',
		description: 'Setup a workflow error to track what is going on here.',
	},
];
const stubs = ['n8n-text', 'n8n-link', 'n8n-icon'];

describe('N8nSuggestedActions', () => {
	it('renders the suggested actions count', () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		expect(wrapper.baseElement).toContainHTML('data-test-id="suggested-action-count"');
		expect(wrapper.getByTestId('suggested-action-count')).toHaveTextContent('0 / 2');
	});

	it('does not render the suggested actions count if all are completed', () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: [
					{ ...mockActions[0], completed: true },
					{ ...mockActions[1], completed: true },
				],
			},
		});

		expect(wrapper.baseElement).not.toContainHTML('data-test-id="suggested-action-count"');
	});

	it('renders the suggested actions count with completed actions', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: [{ ...mockActions[0], completed: true }, mockActions[1]],
			},
			global: { stubs },
		});

		expect(wrapper.getByTestId('suggested-action-count')).toBeInTheDocument();
		expect(wrapper.getByTestId('suggested-action-count')).toHaveTextContent('1 / 2');

		const countTag = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(countTag);

		// Check if action items are visible and checked off
		const actionItems = wrapper.getAllByTestId('suggested-action-item');
		expect(actionItems).toHaveLength(2);
		expect(actionItems).toMatchSnapshot();
	});

	it('opens popover when count tag is clicked and all items are not completed', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
			global: { stubs },
		});

		const countTag = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(countTag);

		// Check if action items are visible
		const actionItems = wrapper.getAllByTestId('suggested-action-item');
		expect(actionItems).toHaveLength(2);
		expect(actionItems).toMatchSnapshot();
	});

	it('renders all action items correctly', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		const countTag = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(countTag);

		// Check first action
		expect(wrapper.getByText('Evaluate your workflow')).toBeInTheDocument();
		expect(
			wrapper.getByText('Set up an AI evaluation to be sure th WF is reliable.'),
		).toBeInTheDocument();

		// Check second action
		expect(wrapper.getByText('Keep track of execution errors')).toBeInTheDocument();
		expect(
			wrapper.getByText('Setup a workflow error to track what is going on here.'),
		).toBeInTheDocument();
	});

	it('emits action-click event when action button is clicked', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		const countTag = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(countTag);

		const actionButtons = wrapper.getAllByTestId('suggested-action-item');
		await fireEvent.click(actionButtons[0]);

		expect(wrapper.emitted('action-click')).toBeTruthy();
		expect(wrapper.emitted('action-click')[0]).toEqual(['action1']);
	});

	it('emits ignore-click event when ignore link is clicked', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		const countTag = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(countTag);

		const ignoreLinks = wrapper.getAllByTestId('suggested-action-ignore');
		await fireEvent.click(ignoreLinks[0]);

		// Wait for the delayed emission
		await new Promise((resolve) => setTimeout(resolve, 600));

		expect(wrapper.emitted('ignore-click')).toBeTruthy();
		expect(wrapper.emitted('ignore-click')[0]).toEqual(['action1']);
	});

	it('exposes openPopover method', async () => {
		const { container } = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		// Check that the component renders correctly
		expect(container.querySelector('[data-test-id="suggested-action-count"]')).toBeInTheDocument();
	});

	it('shows custom ignore all text when ignoreAllLabel is provided', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
				ignoreAllLabel: 'Ignore for all',
			},
			global: { stubs: { N8nIcon: true } },
		});

		const countTag = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(countTag);

		expect(wrapper.getByTestId('suggested-action-ignore-all')).toBeInTheDocument();
		expect(wrapper.getByText('Ignore for all')).toBeInTheDocument();
	});

	it('emits ignore-all event when turn off link is clicked', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
				ignoreAllLabel: 'Ignore for all',
			},
			global: { stubs: { N8nIcon: true } },
		});

		const bellButton = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(bellButton);

		const turnOffLink = wrapper.getByTestId('suggested-action-ignore-all');
		expect(wrapper.getByText('Ignore for all')).toBeInTheDocument();
		await fireEvent.click(turnOffLink);

		expect(wrapper.emitted('ignore-all')).toBeTruthy();
	});

	it('renders more info link when moreInfoLink is provided', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		const bellButton = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(bellButton);

		const moreInfoLinks = wrapper.getAllByText('More info');
		expect(moreInfoLinks).toHaveLength(1); // Only first action has moreInfoLink

		const link = moreInfoLinks[0].closest('a') as HTMLAnchorElement;
		expect(link.getAttribute('href')).toBe('https://docs.n8n.io/evaluations');
	});

	it('applies ignoring class when action is being ignored', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		const bellButton = wrapper.getByTestId('suggested-action-count');
		await fireEvent.click(bellButton);

		const actionItems = wrapper.getAllByTestId('suggested-action-item');
		const ignoreLinks = wrapper.getAllByTestId('suggested-action-ignore');

		// Click ignore on first action
		await fireEvent.click(ignoreLinks[0]);

		// Check if the action item has the ignoring class
		expect(actionItems[0]).toHaveClass('ignoring');
	});

	it('exposes closePopover method', async () => {
		const { container } = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
			},
		});

		// Check that the component renders correctly (closePopover is exposed but we can't test it directly in this environment)
		expect(container.querySelector('[data-test-id="suggested-action-count"]')).toBeInTheDocument();
	});

	it('respects popoverAlignment prop', async () => {
		const wrapper = render(N8nSuggestedActions, {
			props: {
				actions: mockActions,
				popoverAlignment: 'start',
			},
		});

		// Check that the component renders correctly with the alignment prop
		expect(wrapper.getByTestId('suggested-action-count')).toBeInTheDocument();
	});
});
