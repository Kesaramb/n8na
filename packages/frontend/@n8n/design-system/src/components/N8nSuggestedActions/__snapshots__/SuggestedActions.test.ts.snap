// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`N8nSuggestedActions > opens popover when count tag is clicked and all items are not completed 1`] = `
[
  <div
    class="actionItem actionable"
    data-action-id="action1"
    data-test-id="suggested-action-item"
  >
    <div
      class="checkboxContainer"
    >
      <n8n-icon-stub
        color="foreground-dark"
        icon="circle"
        spin="false"
      />
    </div>
    <div
      class="actionItemBody"
    >
      <div
        class="mb-3xs actionHeader"
      >
        <n8n-text-stub
          bold="true"
          compact="false"
          size="medium"
          tag="span"
        />
        <n8n-icon-stub
          icon="chevron-right"
          spin="false"
        />
      </div>
      <div>
        <n8n-text-stub
          bold="false"
          color="text-base"
          compact="false"
          size="small"
          tag="span"
        />
      </div>
      <div
        class="actionButtons"
      >
        <n8n-link-stub
          bold="false"
          data-test-id="suggested-action-ignore"
          newwindow="false"
          size="small"
          theme="text"
          underline="true"
        />
      </div>
    </div>
  </div>,
  <div
    class="actionItem actionable"
    data-action-id="action2"
    data-test-id="suggested-action-item"
  >
    <div
      class="checkboxContainer"
    >
      <n8n-icon-stub
        color="foreground-dark"
        icon="circle"
        spin="false"
      />
    </div>
    <div
      class="actionItemBody"
    >
      <div
        class="mb-3xs actionHeader"
      >
        <n8n-text-stub
          bold="true"
          compact="false"
          size="medium"
          tag="span"
        />
        <n8n-icon-stub
          icon="chevron-right"
          spin="false"
        />
      </div>
      <div>
        <n8n-text-stub
          bold="false"
          color="text-base"
          compact="false"
          size="small"
          tag="span"
        />
      </div>
      <div
        class="actionButtons"
      >
        <n8n-link-stub
          bold="false"
          data-test-id="suggested-action-ignore"
          newwindow="false"
          size="small"
          theme="text"
          underline="true"
        />
      </div>
    </div>
  </div>,
]
`;

exports[`N8nSuggestedActions > renders the suggested actions count with completed actions 1`] = `
[
  <div
    class="actionItem"
    data-action-id="action1"
    data-test-id="suggested-action-item"
  >
    <div
      class="checkboxContainer"
    >
      <n8n-icon-stub
        color="success"
        icon="circle-check"
        spin="false"
      />
    </div>
    <div
      class="actionItemBody"
    >
      <div
        class="actionHeader"
      >
        <n8n-text-stub
          bold="true"
          compact="false"
          size="medium"
          tag="span"
        />
        <!--v-if-->
      </div>
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>,
  <div
    class="actionItem actionable"
    data-action-id="action2"
    data-test-id="suggested-action-item"
  >
    <div
      class="checkboxContainer"
    >
      <n8n-icon-stub
        color="foreground-dark"
        icon="circle"
        spin="false"
      />
    </div>
    <div
      class="actionItemBody"
    >
      <div
        class="mb-3xs actionHeader"
      >
        <n8n-text-stub
          bold="true"
          compact="false"
          size="medium"
          tag="span"
        />
        <n8n-icon-stub
          icon="chevron-right"
          spin="false"
        />
      </div>
      <div>
        <n8n-text-stub
          bold="false"
          color="text-base"
          compact="false"
          size="small"
          tag="span"
        />
      </div>
      <div
        class="actionButtons"
      >
        <n8n-link-stub
          bold="false"
          data-test-id="suggested-action-ignore"
          newwindow="false"
          size="small"
          theme="text"
          underline="true"
        />
      </div>
    </div>
  </div>,
]
`;
